Source/DafnyRuntime/DafnyRuntimeDafny/src/dafnyRuntime.dfy

```diff
@@ -0,0 +1,865 @@
+
+      Repr := {this} + storage.Repr;
+    }
+
+    method RemoveLast() returns (t: T)
+      requires Valid()
+      requires 0 < size
+      modifies Repr
+      ensures ValidAndDisjoint()
+      ensures old(Value()) == Value() + [t]
+      ensures Value() == old(Value()[..(size - 1)])
+      ensures t in old(Value())
+    {
+      t := storage.Select(size - 1);
+      size := size - 1;
+    }
+
+    method Append(other: ImmutableArray<T>)
+      requires Valid()
+      requires other.Valid()
+      requires size as int + other.Length() as int < SIZE_T_LIMIT
+      modifies Repr
+      ensures ValidAndDisjoint()
+      ensures Value() == old(Value()) + other.values
+    {
+      var newSize := size + other.Length();
+      EnsureCapacity(newSize);
+      storage.UpdateSubarray(size, other);
+      size := size + other.Length();
+    }
+
+    method Freeze() returns (ret: ImmutableArray<T>)
+      requires Valid()
+      ensures ret.Valid()
+      ensures ret.values == Value()
+      // Explicitly doesn't ensure Valid()!
+    {
+      ret := storage.Freeze(size);
+    }
+  }
+
+  // A way to ensure mutable state is safe for concurrent reads and writes
+  // without risk of corrupted values.
+  // The external implementation of this is expected to use language features
+  // such as `volatile` as necessary to ensure that partial values are never read.
+  //
+  // Note that it has no mutable state according to Dafny's verification model,
+  // since Put modifies {}! This is safe as long as only Dafny source code calls
+  // the class' methods, since Dafny ensures that all values satisfy the invariant.
+  // But if you Put(t), you don't get to assume that a subsequent Get() will return t.
+  // This accurately models the nondeterminism inherent to data races.
+  //
+  // Compilers could special case this type to inline declarations and avoid
+  // the cost of allocation and indirection, e.g. by replacing a `const fooBox: AtomicBox<T>`
+  // with a direct `volitile T fooBox;` in the target language.
+  trait {:extern} AtomicBox<T> {
+
+    ghost const inv: T -> bool
+
+    ghost predicate Valid()
+
+    static method {:extern} Make(ghost inv: T -> bool, t: T) returns (ret: AtomicBox<T>)
+      requires inv(t)
+      ensures ret.Valid()
+      ensures ret.inv == inv
+
+    method {:extern} Get() returns (t: T)
+      requires Valid()
+      ensures inv(t)
+
+    method {:extern} Put(t: T)
+      requires Valid()
+      requires inv(t)
+  }
+
+  trait {:extern} Sequence<T> {
+
+    // This is only here to support the attempts some runtimes make to
+    // track what sequence values are actually sequences of characters.
+    // This is not used when --unicode-char is enabled.
+    var isString: bool
+
+    // Total class instances in the tree.
+    // Used in decreases clauses.
+    ghost const NodeCount: nat
+
+    ghost predicate Valid()
+      decreases NodeCount, 0
+      ensures Valid() ==> 0 < NodeCount
+
+    function Cardinality(): size_t
```

Comment:
I decided to stick to the terminology used in Dafny documentation like the reference manual as much as possible (and hence using "Select" instead of "Index" for example). The existing Go runtime code uses Cardinality for other collections too.